### ISSUE TITLE
Question: HTTPS and Appium + iOS

### ISSUE THREAD

**Issue by nderevjanik-skillz** [httptoolkit/mockttp#46]
I'm not able to get HTTPS to work. I followed these instructions:
https://github.com/httptoolkit/mockttp/blob/master/docs/setup.md#mocking-https

Installed the generated cert on my iOS simulator (I can see the cert under Certificate Trust Settings) and then configured mockttp to use the cert and key. I've also configured my system's network connection (macOS Catalina 10.15.7) to use the proxy for HTTP and HTTPS traffic (the simulator uses the machine's network configuration).

I'm then launching and interacting with my app under test using Appium (Webdriver.io). However the app fails when making HTTPS connections.

Any suggestions on how to debug this scenario?

**Comment by pimterry** [comment]
What's the error when the app connection fails? Are there any specific details on why the request fails in your logs?

I would:

* Pass `debug: true` as an option to Mockttp's `getLocal` - it enables a bunch of extra logging which might have some clues. You should at least see "Generating certificate for..." while setting up HTTPS connections and "Handling request for..." for each subsequent received request. If you don't see anything, then your device isn't connecting to Mockttp at all.
* Try running `openssl s_client -connect localhost:<port>` against your Mockttp server - this tells OpenSSL to connect to that port and log the certificate used and connection parameters. You can then that a) it's using HTTPS correctly and b) the cert details are as you expect them to be.
* Try using your Mockttp server as a proxy from your desktop browser, which might provide a more detailed explanation for HTTPS issues than your app's HTTP library.

**Comment by nderevjanik-skillz** [comment]
I tried the first two bullets (and forgive me but my knowledge of SSL / TLS is very limited):

```shell
[0-0] Starting mock server on port 8080
[0-0] STEP: Clicking on the Launch button.
[0-0] Proxying HTTP/1 CONNECT to calendar.google.com:443
[0-0] Generating certificate for calendar.google.com
[0-0] TLS client error: {"failureCause":"cert-rejected","hostname":"calendar.google.com","remoteIpAddress":"::1","tags":[]}
[0-0] Proxying HTTP/1 CONNECT to api.staging.company.com:443
[0-0] Generating certificate for api.staging.company.com
[0-0] TLS client error: {"failureCause":"reset","hostname":"api.staging.company.com","remoteIpAddress":"::1","tags":[]}
```

Note: I've redacted the actual hostname (api.staging.company.com) but its similar to this.

```shell
CONNECTED(00000005)
depth=1 CN = Mockttp Testing CA - DO NOT TRUST
verify error:num=19:self signed certificate in certificate chain
verify return:0
---
Certificate chain
 0 s:/CN=localhost/O=Mockttp Cert - DO NOT TRUST
   i:/CN=Mockttp Testing CA - DO NOT TRUST
 1 s:/CN=Mockttp Testing CA - DO NOT TRUST
   i:/CN=Mockttp Testing CA - DO NOT TRUST
---
Server certificate
-----BEGIN CERTIFICATE-----
… redacted …
-----END CERTIFICATE-----
subject=/CN=localhost/O=Mockttp Cert - DO NOT TRUST
issuer=/CN=Mockttp Testing CA - DO NOT TRUST
---
No client certificate CA names sent
Server Temp Key: ECDH, X25519, 253 bits
---
SSL handshake has read 1917 bytes and written 289 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES128-GCM-SHA256
Server public key is 1024 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES128-GCM-SHA256
    Session-ID: … redacted …
    Session-ID-ctx: 
    Master-Key: … redacted …
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    … redacted …

    Start Time: 1610662504
    Timeout   : 7200 (sec)
    Verify return code: 19 (self signed certificate in certificate chain)
---
read:errno=0
```

**Comment by pimterry** [comment]
Ok, that suggests that Mockttp is correctly configured with a CA certificate, and it is receiving the requests from the client so your proxy settings are correct, and the issue is that the client doesn't trust the CA.

It's most likely I think that your client isn't correctly configured somehow. How exactly have you set that up?

**Comment by nderevjanik-skillz** [comment]
I took the testCA.pem generated from:
https://github.com/httptoolkit/mockttp/blob/master/docs/setup.md#generating-a-certificate

And dropped it on the iOS simulator which automatically installs and trusts it:

<img src="https://user-images.githubusercontent.com/73610724/104658714-41817700-5678-11eb-8234-81c4fdc7c66b.png" width="300">

**Comment by pimterry** [comment]
Hmm, that looks pretty convincing. What HTTP client are you using in the app, and how is that configured? I don't do much iOS development nowadays, do you need to explicitly manage certificate trust there too, or opt into using the system certificate store or similar?

**Comment by nderevjanik-skillz** [comment]
I'll need to speak with my development and get back to you.

Btw, I really appreciate you taking the time. I'm existed to get a pure JS/TS solution working.

**Comment by pimterry** [comment]
Hi @nderevjanik-skillz are you still having trouble with this? Let me know if there's anything else I can do to help. And if you've fixed it, I'd love to know how, for the next user who hits the same problem.

**Comment by nderevjanik-skillz** [comment]
I've had to put this on hold.

On a related note I was able to use mitmproxy. Their server creates SSL certs when you first run it. Which I then install on my device under test. I even tried to configure mockttp to use those certs, no luck.

I was really hoping to use a pure JS/TS solution like yours. So I don't want to give up, I just can't focus on it now.

Feel free to close this "issue", if you don't mind me coming back later and re-opening it when I have more time.

**Comment by pimterry** [comment]
@nderevjanik-skillz I suspect I've independently found the cause of this, when investigating HTTPS issues elsewhere: some clients now reject the size of the default keys Mockttp uses for interception.

The default keys generated by Mockttp when generating HTTPS certificates for sites dynamically are 1024 bits, if you don't specify a length yourself. That was OK when that code was written, but many newer client versions now require keys of 2048 bits or longer. I can see the key in your output above is indeed only 1024 bits, and I suspect iOS is also one of the clients where that's no longer sufficient.

I'm going to update the default now, but it's very quick for you to confirm this without that: just add `keyLength: 2048` to the HTTPS options you pass to Mockttp, to explicitly use longer keys.

That should look something like this:

```javascript
const mockServer = getLocal({
    https: {
        keyPath: '...'
        certPath: '...'
        keyLength: 2048 // <--
    }
});
```

You don't need to regenerate your CA key to do this - it's the certificates generated for each site that are too small, not the CA (almost certainly - the openssl command in Mockttp's setup docs defaults to 2048 for all modern versions of OpenSSL already).

No hurry of course, whenever you're back to this, just noting it here now since I ran into a similar issue myself.

**Comment by nderevjanik-skillz** [comment]
This sounds very promising! I'll give it a try in the next 2 weeks, if not sooner and report back. Thanks again!

**Comment by nderevjanik-skillz** [comment]
Adding keyLength fixed it!

**Comment by pimterry** [comment]
Great! Glad to hear it. I've now released v1.0.6, which uses 2048 as the default key size, so that should solve automatically this in future. If you upgrade you can remove the explicit key length if you prefer, but it doesn't make much difference either way.

I'll close this now, but do let me know if you run into any other issues, and thanks for reporting this!
